src/stark.rs

```diff
@@ -0,0 +1,132 @@
+///
+/// We decided not to implement it due to its cons:
+/// * No support for non-hardened derivation
+/// * Grinding is a probabilistic algorithm which does a lot of hashing (32 hashes
+///   on average, but in worst case can be 1000+).
+/// * In general, it's strange to derive secp256k1 key and then convert it to stark key
+///
+/// Our derivation algorithm addresses these flaws: it yields a stark key right away (without
+/// any intermediate secp256k1 keys), supports non-hardened derivation, does only 2 hashes per
+/// derivation.
+///
+/// [argent-x]: https://github.com/argentlabs/argent-x/blob/13142607d83fea10b297d6a23452e810605784d1/packages/extension/src/shared/signer/ArgentSigner.ts#L14-L25,
+pub struct Stark;
+
+impl DeriveShift<curves::Stark> for Stark {
+    fn derive_public_shift(
+        parent_public_key: &ExtendedPublicKey<curves::Stark>,
+        child_index: NonHardenedIndex,
+    ) -> DerivedShift<curves::Stark> {
+        let hmac = HmacSha512::new_from_slice(&parent_public_key.chain_code)
+            .expect("this never fails: hmac can handle keys of any size");
+        let i0 = hmac
+            .clone()
+            .chain_update([0x00])
+            .chain_update(parent_public_key.public_key.to_bytes(true))
+            .chain_update(child_index.to_be_bytes())
+            .finalize()
+            .into_bytes();
+        let i1 = hmac
+            .chain_update([0x01])
+            .chain_update(parent_public_key.public_key.to_bytes(true))
+            .chain_update(child_index.to_be_bytes())
+            .finalize()
+            .into_bytes();
+        Self::calculate_shift(parent_public_key, i0, i1)
+    }
+
+    fn derive_hardened_shift(
+        parent_key: &ExtendedKeyPair<curves::Stark>,
+        child_index: HardenedIndex,
+    ) -> DerivedShift<curves::Stark> {
+        let hmac = HmacSha512::new_from_slice(parent_key.chain_code())
+            .expect("this never fails: hmac can handle keys of any size");
+        let i0 = hmac
+            .clone()
+            .chain_update([0x00])
+            .chain_update([0x00])
+            .chain_update(parent_key.secret_key.secret_key.as_ref().to_be_bytes())
+            .chain_update(child_index.to_be_bytes())
+            .finalize()
+            .into_bytes();
+        let i1 = hmac
+            .chain_update([0x01])
+            .chain_update([0x00])
+            .chain_update(parent_key.secret_key.secret_key.as_ref().to_be_bytes())
+            .chain_update(child_index.to_be_bytes())
+            .finalize()
+            .into_bytes();
+        Self::calculate_shift(&parent_key.public_key, i0, i1)
+    }
+}
+
+impl Stark {
+    fn calculate_shift(
+        parent_public_key: &ExtendedPublicKey<curves::Stark>,
+        i0: hmac::digest::Output<HmacSha512>,
+        i1: hmac::digest::Output<HmacSha512>,
+    ) -> DerivedShift<curves::Stark> {
+        let i = generic_array::sequence::Concat::concat(i0, i1);
+        let (shift, chain_code) = split(&i);
```

Comment:
Wait, why is concatenation not done in the calling function? Does this optimize better?

Reply:
To have less of repeating code. Two different function are calling this one, if we were doing concat in the calling functions, we needed to write concat twice.

Reply:
But this way you leak implementation details! It's not a big problem since it's an internal function, but it looks like a bizzare choice to me to mangle the domains like this to save one (1) line

Reply:
Since it's an internal function, it's not leaking implementation detail

Reply:
I dunno, it feels like a natural choice for me...